### PR TITLE
feat(facets-next): Add `FacetsService`

### PR DESCRIPTION
**Pull request by javieri-empathy**
# Description

Adds the facets service. This new service receives DTOs, transforms them to entities, and performs operations with them.

# Context

This is needed for the new facets module. The wiring will use this service, instead of invoking actions or mutations.

# Test

Pay special attention to the provided tests, check that they are clear enough, and if there are any missing use cases.

### CODE REVIEW

File: packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+      expect(isFilterSelected(sizeFilter)).toBe(true);
+      expect(isFilterSelected(priceFilter)).toBe(true);
+      expect(isEditableNumberRangeFilterSelected(ageFilter as EditableNumberRangeFilter)).toBe(
+        true
+      );
+    });
+  });
+  describe('saves a group of facets', () => {
+    it('saves groups of facets keeping its previous selected states', () => {
+      const { service, getSelectedFilters, getFilters, getStoreEditableNumberRangeFilter } =
```
Comment (javieri-empathy) [inline packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts:314]: I'd like to have some time to rethink this test. It represents a real use case, but it is too verbose.

File: packages/x-components/src/x-modules/facets-next/facets-service.types.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * The facet group is used to perform operations over a list of facets. For example, some APIs
+ * have dynamic facets. This means that depending on the selected filters, some facets can be
+ * returned. For example if you select `category:television`, a new facet called *Screen size* might
+ * appear. And, because now filters are stored in a dictionary instead of an array, and overridden
+ * with each request, when the user performs a new query, we have to remove all the `Screen size`
+ * filters because we don't know if the API will return them or not in the new query.
+ *
+ * @public
+ */
+export interface FacetGroup {
```
Comment (javieri-empathy) [inline packages/x-components/src/x-modules/facets-next/facets-service.types.ts:15]: Consider moving this to the search-types package.

File: packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts [hunk 3]
```diff
@@ -42,8 +42,10 @@ export class HierarchicalFilterEntity implements FilterEntity {
     if (filter.children) {
       filter.children.forEach(childId => {
         const child = this.getFilterById(childId);
-        this.saveFilter({ ...child, selected: false });
-        this.deselectDescendants(child);
+        if (child) {
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts:45]: adding test for this case?

Comment (javieri-empathy) [inline packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts:45]: There is a problem with the current approach, and it is that a filter child might not be (yet) in the store. We could maybe add them always before using an entity (at the service), or pass to the entities the list of filters they should work with so they do not have to rely on querying the store.

File: packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+      service.clearFilters(['category', 'color']);
+      expect(getSelectedFilters()).toHaveLength(3);
+      expect(isFilterSelected(sizeFilter)).toBe(true);
+      expect(isFilterSelected(priceFilter)).toBe(true);
+      expect(isEditableNumberRangeFilterSelected(ageFilter as EditableNumberRangeFilter)).toBe(
+        true
+      );
+    });
+  });
+  describe('saves a group of facets', () => {
```
Comment (pmareke) [inline packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts:312]: Add a blank line here, please

File: packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+          redColorFilter,
+          menCategoryFilter,
+          age10To18Filter
+        ])
+      ).toBe(false);
+    });
+  });
+});
+
+interface FacetsServiceTestAPI {
```
Comment (pmareke) [inline packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts:442]: Please comment this interface.

File: packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+/**
+ * Creates a fresh new {@link BaseFacetsService} with some helpful test methods.
+ *
+ * @returns An object containing methods for testing {@link BaseFacetsService}.
+ */
+function prepareFacetsService(): FacetsServiceTestAPI {
+  XPlugin.resetInstance();
+  installNewXPlugin();
+  XPlugin.registerXModule(facetsNextXModule);
+  return {
```
Comment (pmareke) [inline packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts:36]: Add blank line here please.

**Review by pmareke — CHANGES_REQUESTED** [review]
Great, great job!

File: packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts (hunk 1 above)
Comment (tajespasarela) [reply packages/x-components/src/x-modules/facets-next/__tests__/facets-service.spec.ts:314]: It's ok as it is. Long and verbose? yes, does it test correctly? yes

let's move on!

**Review by tajespasarela — CHANGES_REQUESTED** [review]

File: packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts (hunk 3 above)
Comment (tajespasarela) [reply packages/x-components/src/x-modules/facets-next/entities/hierarchical-filter.entity.ts:45]: it is ok as it is at the moment.

File: packages/x-components/src/x-modules/facets-next/facets-service.types.ts (hunk 2 above)
Comment (javieri-empathy) [reply packages/x-components/src/x-modules/facets-next/facets-service.types.ts:15]: It shouldn't, I remember discussing several ways to set facets groups, but none of them was creating `FacetGroup` objects manually from TypeScript code.

File: packages/x-components/src/x-modules/facets-next/facets.service.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      this.saveFilters(facet.filters, previousFilters);
+    });
+  }
+
+  select(filter: Filter): void {
+    this.createEntity(filter).select(filter);
+  }
+
+  toggle(filter: Filter): void {
+    if (filter.selected) {
```
[the remark this answers is not in the document]
Comment (javieri-empathy) [reply packages/x-components/src/x-modules/facets-next/facets.service.ts:54]: A ternary is an expression, not a statement, so we shouldn't use it there.

File: packages/x-components/src/x-modules/facets-next/facets.service.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+   *
+   * @param filter - The filter to remove.
+   * @internal
+   */
+  protected removeFilter(filter: Filter): void {
+    this.store.commit('x/facetsNext/removeFilter', filter);
+  }
+
+  /**
+   * Removes the filters that belong to the given group.
```
[the remark this answers is not in the document]
Comment (javieri-empathy) [reply packages/x-components/src/x-modules/facets-next/facets.service.ts:93]: It makes no sense to say that. The store does not belong to any group. There is a list of facets that belongs to a group, and filters belongs to a facet. This method removes the filters that belong to a facet that belongs to the given group.

**Review by pmareke — APPROVED** [review]

**Review by pmareke — APPROVED** [review]

**Review by tajespasarela — APPROVED** [review]

**Pull request merged**